code/game/machinery/doors/firedoor.dm

```diff
@@ -203,30 +199,99 @@
+	for(var/obj/machinery/door/firedoor/firelock as anything in merge_group.members)
+		firelock.issue_turfs = shared_problems
+		for(var/dir in GLOB.cardinals)
+			var/turf/checked_turf = get_step(get_turf(firelock), dir)
+			if(!checked_turf)
+				continue
+			process_results(checked_turf)
+
+/obj/machinery/door/firedoor/proc/register_adjacent_turfs(atom/loc)
+	for(var/dir in GLOB.cardinals)
+		var/turf/checked_turf = get_step(get_turf(loc), dir)
+
+		if(!checked_turf)
+			continue
+		process_results(checked_turf)
+		RegisterSignal(checked_turf, COMSIG_TURF_EXPOSE, .proc/process_results)
+		RegisterSignal(checked_turf, COMSIG_TURF_CALCULATED_ADJACENT_ATMOS, .proc/process_results)
+
+
+/obj/machinery/door/firedoor/proc/unregister_adjacent_turfs(atom/loc)
+	for(var/dir in GLOB.cardinals)
+		var/turf/checked_turf = get_step(get_turf(loc), dir)
+	
+		if(!checked_turf)
+			continue
+
+		UnregisterSignal(checked_turf, COMSIG_TURF_EXPOSE)
+		UnregisterSignal(checked_turf, COMSIG_TURF_CALCULATED_ADJACENT_ATMOS)
+
+/obj/machinery/door/firedoor/proc/check_atmos(turf/checked_turf)
+	var/datum/gas_mixture/environment = checked_turf.return_air()
+	var/obj/machinery/airalarm/air_alarm = LAZYACCESS(my_area.airalarms,1)
+	
+	if(air_alarm)
+		var/datum/tlv/selected_temperature = air_alarm.TLV["temperature"]
+		if(environment?.temperature >= selected_temperature?.hazard_max)
+			return FIRELOCK_ALARM_TYPE_HOT
+		if(environment?.temperature <= selected_temperature?.hazard_min)
+			return FIRELOCK_ALARM_TYPE_COLD
+	else
+		if(environment?.temperature >= FIRE_MINIMUM_TEMPERATURE_TO_EXIST)
+			return FIRELOCK_ALARM_TYPE_HOT
+		if(environment?.temperature <= BODYTEMP_COLD_DAMAGE_LIMIT)
+			return FIRELOCK_ALARM_TYPE_COLD
+	return
+
+/obj/machinery/door/firedoor/proc/process_results(datum/source)
+	SIGNAL_HANDLER
+
 	if(!COOLDOWN_FINISHED(src, detect_cooldown))
 		return
-	if(alarm_type)
+	if(alarm_type == FIRELOCK_ALARM_TYPE_GENERIC)
 		return
+
 	for(var/area/place in affecting_areas)
 		if(!place.fire_detect) //if any area is set to disable detection
 			return
 
-	var/turf/my_turf = source
-	if(!my_turf)
-		my_turf = get_turf(src)
-
-	var/datum/gas_mixture/environment = my_turf.return_air()
-	var/result
-
-	if(environment?.temperature >= FIRE_MINIMUM_TEMPERATURE_TO_EXIST)
-		result = FIRELOCK_ALARM_TYPE_HOT
-	if(environment?.temperature <= BODYTEMP_COLD_DAMAGE_LIMIT)
-		result = FIRELOCK_ALARM_TYPE_COLD
-	if(!result)
+	var/turf/checked_turf = source
+	var/result = check_atmos(checked_turf)
+	
+	if(!length(issue_turfs) && alarm_type)
+		start_deactivation_process()
 		return
-
-	start_activation_process(result)
+	else if(length(issue_turfs) && !alarm_type)
+		result = check_atmos(issue_turfs[1])
+		if(result)
+			start_activation_process(result)
+			return
+	if(result && TURF_SHARES(checked_turf) && issue_turfs)
+		issue_turfs |= checked_turf
+	else if((!result && length(issue_turfs) || !TURF_SHARES(checked_turf)) && issue_turfs)
+		issue_turfs -= checked_turf
```

Comment:
I feel like you're running into trouble with the order you're doing this in, and it's placing you in a position where you need to do some kinda weird things. Just now noticed.

The code gets a result, then attempts to divine if the result would lead to a closure or not. Then it adds/removes.
I feel like you'd have a better time if you added/removed, and then activated or deactivated based off the state then. So something like this

```
store the current length of the issue turf list

check to see if the turf can share/its alarm is a valid reason to activate
if it is, add the turf, then check the old length. If it's `> 0 [if(!old_issue_len) -> return]`, do nothing, if it's not, activate the firelocks

If the turf isn't valid to activate, remove it from the issue turfs list. Then check if the issue turfs list is now totally empty. 
if it is, deactivate.
```

Reply:
did some cleanup